stdlib/public/runtime/ProtocolConformance.cpp

```diff
@@ -332,6 +333,31 @@ static int _addImageProtocolConformances(struct dl_phdr_info *info,
   dlclose(handle);
   return 0;
 }
+#elif defined(__CYGWIN__)
+static int _addImageProtocolConformances(struct dl_phdr_info *info,
```

Comment:
In my opnion, there are three kinds of platform specific implementation in performing '_initializeCallbacksToInspectDylib'.

First, the function for seaching loaded dls/images/DLLs is diffrent for every platform.
The function is different for every platform - _dyld_register_func_for_add_image(), dl_iterate_phdr(), android_iterate_libs(), _swift_dl_iterate_phdr (in CYGWIN). Some of these are supported by system library( APPLE , Linux) and others are implemented for swift.
For that reason, there are conditional preprocessing in two '_initializeCallbacksToInspectDylib'.

Second, the callback function which calling by the searching function is different.
Although the callback function name is same for each platform, the prototype is different. The reason why is because the different searching function specifies different prototype of callback.

Finally, the method for inspecting a section of dl/image/DLL is different.
This method is directly supported by system library (getsectiondata) in APPLE, or implemented with POSIX dlsym and linker script in Linux/Android, or implemented for swift in Cygwin. Except APPLE, they all use POSIX dlopen/dlclose for handling  dl/image/DLL.

I think it is most important to generalize two '_initializeCallbacksToInspectDylib' from where the duplication is begining for every platform.

How about introducing general function _swift_inspectDylibs as follows?

a sketch : [unify.cpp](https://github.com/tinysun212/swift-cygwin/blob/458965d2b70b24638e24d21ec00f5001f802c3bb/misc/unify.cpp)